frontend/src/scenes/data-warehouse/editor/sidebar/UpstreamGraph.tsx

```diff
@@ -0,0 +1,232 @@
+    dagreGraph.setGraph({ rankdir: 'LR', nodesep: 68, ranksep: 160 })
+
+    // Add nodes and edges to dagre
+    nodes.forEach((node) => {
+        const isMatView = node.type === 'view' && !!node.last_run_at
+        dagreGraph.setNode(node.id, { width: 240, height: isMatView ? MAT_VIEW_HEIGHT : TABLE_HEIGHT })
+    })
+    edges.forEach((edge) => {
+        dagreGraph.setEdge(edge.source, edge.target)
+    })
+
+    // Calculate layout
+    dagre.layout(dagreGraph)
+
+    // Convert back to ReactFlow format
+    const layoutedNodes: Node[] = nodes.map((node) => {
+        const nodeWithPosition = dagreGraph.node(node.id)
+        const isMatView = node.type === 'view' && !!node.last_run_at
+        return {
+            id: node.id,
+            type: 'lineageNode',
+            position: {
+                x: nodeWithPosition.x - nodeWithPosition.width / 2,
+                y: nodeWithPosition.y - nodeWithPosition.height / 2,
+            },
+            data: {
+                ...node,
+                isCurrentView: node.name === currentViewName,
+            },
+            width: 240,
+            height: isMatView ? MAT_VIEW_HEIGHT : TABLE_HEIGHT,
+        }
+    })
+
+    const layoutedEdges: Edge[] = edges.map((edge, index) => ({
+        id: `edge-${index}`,
+        source: edge.source,
+        target: edge.target,
+        type: 'smoothstep',
+        animated: false,
+        markerEnd: {
+            type: MarkerType.ArrowClosed,
+            width: 20,
+            height: 20,
+        },
+    }))
+
+    return { nodes: layoutedNodes, edges: layoutedEdges }
+}
+
+function UpstreamGraphContent({ codeEditorKey }: UpstreamGraphProps): JSX.Element {
+    const { upstream, editingView } = useValues(multitabEditorLogic({ key: codeEditorKey }))
+    const { fitView } = useReactFlow()
+    const { isDarkModeOn } = useValues(themeLogic)
+
+    const { nodes, edges } = useMemo(() => {
+        if (!upstream || upstream.nodes.length === 0) {
+            return { nodes: [], edges: [] }
+        }
+
+        return getLayoutedElements(upstream.nodes, upstream.edges, editingView?.name)
+    }, [upstream, editingView?.name])
+
+    // Fit view when nodes change
+    useEffect(() => {
+        if (nodes.length > 0) {
+            setTimeout(() => fitView({ padding: 0.1 }), 100)
+        }
+    }, [nodes, fitView])
+
+    if (!upstream || upstream.nodes.length === 0) {
+        return (
+            <div className="w-full h-full flex items-center justify-center text-muted">
+                <div className="text-center">
+                    <div className="text-lg mb-2">No tables or views found</div>
+                    <div className="text-sm">This query doesn't depend on any other tables or views</div>
+                </div>
+            </div>
+        )
+    }
+
+    return (
+        <div className="w-full h-full">
+            <ReactFlow
+                proOptions={{
+                    hideAttribution: true,
+                }}
```

Comment:
It wasn't clear to me if we currently pay for this license, but it looks like we may have at least in the past.